api/v1beta1/cluster_types.go

```diff
@@ -669,17 +672,20 @@ type MachineDeploymentTopology struct {
 	// class is the name of the MachineDeploymentClass used to create the set of worker nodes.
 	// This should match one of the deployment classes defined in the ClusterClass object
 	// mentioned in the `Cluster.Spec.Class` field.
+	// +kubebuilder:validation:MaxLength=256
 	Class string `json:"class"`
 
 	// name is the unique identifier for this MachineDeploymentTopology.
 	// The value is used with other unique identifiers to create a MachineDeployment's Name
 	// (e.g. cluster's name, etc). In case the name is greater than the allowed maximum length,
 	// the values are hashed together.
+	// +kubebuilder:validation:MaxLength=256
```

Comment:
This is a subset of a CRD name, so definitely can't be longer than 253 chars

Reply:
Yes and no. This gets handed over to a name generator: https://github.com/kubernetes-sigs/cluster-api/blob/3a8728f0e3d25091faf64a9bf92e0eb0bb49065e/exp/topology/desiredstate/desired_state.go#L791

which then cuts it to max 63  (and ensures it ends with a random suffix)

While I would say more than 58 doesn't really make sense, we support it today and cut it down to something that works.

So how safe it is to restrict this field now?

Reply:
If our minimum supported K8s version is 1.30, then validation ratcheting is enabled by default, and we could add an accurate max length, but allow existing values to continue to be valid

Reply:
Sort of. Ratcheting still breaks backup / restore. Also doesn't help if higher level tools have been build under the assumption that they can use names like this and then they break when we add this limitation.

Also in general folks could intentionally use pretty long names here without the expectation that the full string is actually contained in the MD name (as with ClusterClass they can control everything with Cluster.spec.topology and don't have to modify individual MDs). So this could be some sort of pretty name for some folks

Reply:
Ack, can leave as 256 for now then if you think that's a valid use case